perf/bps_tree.cc

```diff
@@ -0,0 +1,464 @@
+
+generate_benchmark_height(tree_i64, delete_insert_rand, 1);
+generate_benchmark_height(tree_i64, delete_insert_rand, 2);
+generate_benchmark_height(tree_i64, delete_insert_rand, 3);
+generate_benchmark_height(tree_i64, delete_insert_rand, 4);
+
+/*
+ * The following functions test insertion and deletion routines including the
+ * reballancing process overhead. The iteration count is specified explicitly
+ * for insertion testing functions in order to allocate enough memory for the
+ * test tree. The iteration count of deletion testing functions is specified
+ * too in order to create trees big enough to perform max possible deletions.
+ */
+
+template<class tree, class KeyGen>
+static void
+test_insert(benchmark::State &state, size_t count, KeyGen kg)
+{
+	typename tree::Allocator allocator(count);
+
+	typename tree::tree_t t;
+	tree::create(&t, 0, allocator.alloc, allocator.free,
+		     &allocator, NULL);
+	typename tree::elem_t replaced, successor;
+	for (auto _ : state)
+		tree::insert(&t, kg(), &replaced, &successor);
+	tree::destroy(&t);
+}
+
+template<class tree>
+static void
+test_insert_first(benchmark::State &state, size_t count)
+{
+	test_insert<tree>(state, count, DecrementingKey(count));
+}
+
+generate_benchmark_size(tree_i64, insert_first, 10000000)->Iterations(10000000);
+
+template<class tree>
+static void
+test_insert_last(benchmark::State &state, size_t count)
+{
+	test_insert<tree>(state, count, IncrementingKey());
+}
+
+generate_benchmark_size(tree_i64, insert_last, 10000000)->Iterations(10000000);
+
+template<class tree>
+static void
+test_insert_rand(benchmark::State &state, size_t count)
+{
+	test_insert<tree>(state, count, RandomKey(count));
+}
+
+generate_benchmark_size(tree_i64, insert_rand, 10000000)->Iterations(10000000);
+
+template<class tree, class KeyGen>
+static void
+test_delete(benchmark::State &state, size_t count, KeyGen kg)
+{
+	std::vector<typename tree::elem_t> arr;
+	for (size_t i = 0; i < count; i++)
+		arr.emplace_back(i);
+
+	typename tree::Allocator allocator(count);
+
+	typename tree::tree_t t;
+	tree::create(&t, 0, allocator.alloc, allocator.free,
+		     &allocator, NULL);
+	tree::build(&t, &arr.front(), count);
+	for (auto _ : state)
+		tree::delete_(&t, kg());
+	tree::destroy(&t);
+}
+
+template<class tree>
+static void
+test_delete_first(benchmark::State &state, size_t count)
+{
+	test_delete<tree>(state, count, IncrementingKey());
+}
+
+generate_benchmark_size(tree_i64, delete_first, 10000000)->Iterations(10000000);
+
+template<class tree>
+static void
+test_delete_last(benchmark::State &state, size_t count)
+{
+	test_delete<tree>(state, count, DecrementingKey(count));
+}
+
+generate_benchmark_size(tree_i64, delete_last, 10000000)->Iterations(10000000);
+
+template<class tree>
+static void
+test_delete_rand(benchmark::State &state, size_t count)
+{
+	test_delete<tree>(state, count, RandomKey(count));
+}
+
+generate_benchmark_size(tree_i64, delete_rand, 10000000)->Iterations(10000000);
+
```

Comment:
What about mixing insertions and deletions? Let's say, randomly or by some pattern.

Reply:
Given we have random insertion and deletion tests, it seems this one will be redundant. Not sure, how should it be interpreted.

Reply:
We test insertions and deletions separately, but in in a practical workload we usually have them interleaved with some patterns (let's say, random blocks get deleted or inserted completely).

Not sure I am making a good point here, so feel free to resolve at your own discretion.

Reply:
> The thing is, when making a benchmark I usually check if it satisfies the following requirements:
> 1. Something (a suspect) can cause a regression in this test without causing regressions in the existing ones.
> 2. The suspect is defined.

Thanks for elaborating! Sounds reasonable.